src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -345,42 +347,53 @@ class WorkspaceService(protected val userInfo: UserInfo, dataSource: DataSource,
 
           } else {
+            val userServiceRef = context.actorOf(UserService.props(UserService.constructor(dataSource, gcsDAO, containerDAO, null), userInfo))
             val groupsByLevel: Map[WorkspaceAccessLevel, Map[Either[RawlsUser, RawlsGroup], WorkspaceAccessLevel]] = updateMap.groupBy({ case (key, value) => value })
-            //go through the access level groups on the workspace and update them
-            workspaceContext.workspace.accessLevels.foreach { case (level, groupRef) =>
-              withRawlsGroup(groupRef, txn) { group =>
-                //remove existing records for users and groups in the acl update list
-                val users = group.users.filter( userRef => !allTheRefs.contains(userRef) )
-                val groups = group.subGroups.filter( groupRef => !allTheRefs.contains(groupRef) )
-
-                //generate the list of new references
-                val newusers = groupsByLevel.getOrElse(level, Map.empty).keys.collect({ case Left(ru) => RawlsUser.toRef(ru) })
-                val newgroups = groupsByLevel.getOrElse(level, Map.empty).keys.collect({ case Right(rg) => RawlsGroup.toRef(rg) })
-
-                containerDAO.authDAO.saveGroup(group.copy( users = users ++ newusers, subGroups = groups ++ newgroups ) ,txn)
-              }
-            }
 
-            val emailNotFoundReports = (aclUpdates.map( wau => wau.email ) diff updateMap.keys.map({
-              case Left(rawlsUser:RawlsUser) => rawlsUser.userEmail.value
-              case Right(rawlsGroup:RawlsGroup) => rawlsGroup.groupEmail.value
-            }).toSeq)
-              .map( email => ErrorReport( StatusCodes.NotFound, email ) )
+            // go through the access level groups on the workspace and update them
+            // foldLeft to go through the futures in order instead of parallel to avoid concurrent modification exceptions
+            workspaceContext.workspace.accessLevels.foldLeft(Future.successful(RequestComplete(StatusCodes.NoContent)).asInstanceOf[Future[PerRequestMessage]]) {
+              case (priorFuture, (level, groupRef)) => priorFuture.flatMap {
+                case RequestComplete(StatusCodes.NoContent) => withRawlsGroup(groupRef, txn) { group =>
+                  //remove existing records for users and groups in the acl update list
+                  val users = group.users.filter(userRef => !allTheRefs.contains(userRef))
+                  val groups = group.subGroups.filter(groupRef => !allTheRefs.contains(groupRef))
+
+                  //generate the list of new references
+                  val newusers = groupsByLevel.getOrElse(level, Map.empty).keys.collect({ case Left(ru) => RawlsUser.toRef(ru) })
+                  val newgroups = groupsByLevel.getOrElse(level, Map.empty).keys.collect({ case Right(rg) => RawlsGroup.toRef(rg) })
+
+                  implicit val timeout = Timeout(1 minutes)
+                  userServiceRef ? UserService.OverwriteGroupMembers(group, RawlsGroupMemberList(
```

Comment:
here is the call to UserService. It produces an untyped Future which we have to cast :(